src/app/config.rs

```diff
@@ -786,6 +642,171 @@ fn parse_dns_suffix(s: &str) -> Result<dns::Suffix, ParseError> {
+}
+
+pub fn parse_identity_config<S: Strings>(strings: &S) -> Result<Option<identity::Config>, Error> {
+    let sa = parse_control_addr(strings, ENV_IDENTITY_SVC_BASE);
+    let ta = parse(strings, ENV_IDENTITY_TRUST_ANCHORS, |ref s| {
+        identity::TrustAnchors::from_pem(s).ok_or(ParseError::InvalidTrustAnchors)
+    });
+    let dir = parse(strings, ENV_IDENTITY_DIR, |ref s| Ok(PathBuf::from(s)));
+    let tok = parse(strings, ENV_IDENTITY_TOKEN_FILE, |ref s| {
+        identity::TokenSource::if_nonempty_file(s.to_string()).map_err(|e| {
+            error!("Could not read {}: {}", ENV_IDENTITY_TOKEN_FILE, e);
+            ParseError::InvalidTokenSource
+        })
+    });
+    let li = parse(strings, ENV_IDENTITY_IDENTITY_LOCAL_NAME, parse_identity);
+    let min_refresh = parse(strings, ENV_IDENTITY_MIN_REFRESH, parse_duration);
+    let max_refresh = parse(strings, ENV_IDENTITY_MAX_REFRESH, parse_duration);
+
+    let disabled = strings
+        .get(ENV_IDENTITY_DISABLED)?
+        .map(|d| !d.is_empty())
+        .unwrap_or(false);
+
+    match (
+        disabled,
+        sa?,
+        ta?,
+        dir?,
+        li?,
+        tok?,
+        min_refresh?,
+        max_refresh?,
+    ) {
+        (disabled, None, None, None, None, None, None, None) => {
+            if !disabled {
+                error!(
+                    "{} must be set or identity configuration must be specified.",
+                    ENV_IDENTITY_DISABLED
+                );
+                Err(Error::InvalidEnvVar)
+            } else {
+                Ok(None)
+            }
+        }
+        (
+            false,
+            Some(svc),
+            Some(trust_anchors),
+            Some(dir),
+            Some(local_name),
+            Some(token),
+            min_refresh,
+            max_refresh,
+        ) => {
+            let key = {
+                let mut p = dir.clone();
+                p.push("key");
+                p.set_extension("p8");
+
+                fs::read(p)
+                    .map_err(|e| {
+                        error!("Failed to read key: {}", e);
+                        Error::InvalidEnvVar
+                    })
+                    .and_then(|b| {
+                        identity::Key::from_pkcs8(&b).map_err(|e| {
+                            error!("Invalid key: {}", e);
+                            Error::InvalidEnvVar
+                        })
+                    })
+            };
+
+            let csr = {
+                let mut p = dir;
+                p.push("csr");
+                p.set_extension("der");
+
+                fs::read(p)
+                    .map_err(|e| {
+                        error!("Failed to read Csr: {}", e);
+                        Error::InvalidEnvVar
+                    })
+                    .and_then(|b| {
+                        identity::Csr::from_der(b).ok_or_else(|| {
+                            error!("No Csr found");
```

Comment:
In the logs, the acronym should probably still be upper-case...I'm assuming this is due to a find and replace?
```suggestion
                            error!("No CSR found");
```